### DISCUSSION TITLE
Is Fragment File alone sufficient as input for Signac?

### DISCUSSION THREAD

**Discussion by lwtan90** [stuart-lab/signac#908] [Q&A] [answered]
Hi,

I would like to reanalyze a published scATAC-seq dataset but the only data provided was fragment file. I am wondering if this is sufficient as an input for Signac? If yes, can you point me to the correct function to do that?
Thank you.

Wilson

**Comment by timoast** [comment, marked as answer]
Yes, you can run a full analysis starting from a fragment file as input. The steps you would need to follow are:

1. Define genomic regions of interest. This can be done using the `CallPeaks()` function to find peaks, or using `AggregateTiles()` or `GenomeBinMatrix()` to quantify genome bins.
2. Quantify regions using `FeatureMatrix()` function if using peaks rather than bins.
3. Create chromatin assay from the quantified counts matrix using `CreateChromatinAssay()`
4. Create Seurat object from the chromatin assay using `CreateSeuratObject()`

**Reply by obenno** [reply]
Hi,

Thanks for the instruction on all the steps. I noticed there is one piece missed. With the "raw" fragments file loaded, all of the barcodes were included. It was mentioned in the paper of `Signac` that it could identify cells from background noncell-containing barcodes. Could you please indicate the function to achieve this? Thank you in advance!

**Reply by timoast** [reply]
You can use the `CountFragments()` function to get the total number of fragment counts per cell barcode: https://satijalab.org/signac/reference/countfragments

**Reply by obenno** [reply]
Hi,

Thank you very much for your prompt reply. I should read the document more carefully, the functions were actually included in the last section of 10x PBMC tutorial. 

From my understanding, the "cells" were actually selected by filtering all barcodes according to the criteria below:

- `min.cells` option of the `CreateChromatinAssay` function
- `min.features` option of the `CreateChromatinAssay` function
- fraction of reads in peaks (FRiP)
- raw fragment counts in peaks
- TSS enrichment score
- fraction of fragments in blacklist region
- nucleosome signal score

I would like to ask does `Signac` have a function to conduct "cell-calling" method as 10x, which is described [here](https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/algorithms/overview#cells)? And it will be the most appreciated if you could kindly confirm the code below is a good practice to analyze non-10x data with `Signac`

<details>
<summary>signac codes</summary>

```
#! /usr/bin/env Rscript

library(dplyr)
library(tibble)
library(Signac)
library(Seurat)
library(GenomeInfoDb)
library(EnsDb.Hsapiens.v75)
library(ggplot2)
library(patchwork)
library(future)

set.seed(1234)

plan("multicore", workers = 10)
options(future.globals.maxSize = 50 * 1024 ^ 3) # for 50 Gb RAM

frag <- CreateFragmentObject("atac_v1_pbmc_10k_fragments.tsv.gz")

macs_peaks <- CallPeaks(frag, macs2.path="/home/xzx/Tools/miniconda3/envs/scATAC/bin/macs2")

counts <- FeatureMatrix(fragments = frag, features = macs_peaks, sep = c(":", "-"))

chrom_assay <- CreateChromatinAssay(
  counts = counts,
  sep = c(":", "-"),
  genome = "hg19",
  fragments = "atac_v1_pbmc_10k_fragments.tsv.gz",
  min.cells = 10,
  min.features = 200
)

pbmc <- CreateSeuratObject(
  counts = chrom_assay,
  assay = "peaks"
)

annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v75)
seqlevelsStyle(annotations) <- 'UCSC'
Annotation(pbmc) <- annotations

# compute nucleosome signal score per cell
pbmc <- NucleosomeSignal(object = pbmc)

# compute TSS enrichment score per cell
pbmc <- TSSEnrichment(object = pbmc, fast = FALSE)

pbmc$high.tss <- ifelse(pbmc$TSS.enrichment > 2, 'High', 'Low')
##TSSPlot(pbmc, group.by = 'high.tss') + NoLegend()

pbmc$nucleosome_group <- ifelse(pbmc$nucleosome_signal > 4, 'NS > 4', 'NS < 4')
##FragmentHistogram(object = pbmc, group.by = 'nucleosome_group')

## Calculate fragments in peaks for each cell
total_fragments <- CountFragments("atac_v1_pbmc_10k_fragments.tsv.gz") %>%
    dplyr::filter(CB %in% colnames(pbmc)) %>% dplyr::select(CB, frequency_count)

pbmc@meta.data <- pbmc@meta.data %>%
    rownames_to_column("CB") %>%
    left_join(total_fragments, by="CB") %>%
    dplyr::rename("fragments" = frequency_count) %>%
    column_to_rownames("CB")
##pbmc$fragments <- total_fragments[colnames(pbmc), "frequency_count"]
pbmc <- FRiP(
  object = pbmc,
  assay = 'peaks',
  total.fragments = 'fragments'
)

## Counting fragments in blacklist region
pbmc$blacklist_fraction <- FractionCountsInRegion(
  object = pbmc,
  assay = 'peaks',
  regions = blacklist_hg19
)

VlnPlot(
  object = pbmc,
  features = c('FRiP', 'nCount_peaks',
               'TSS.enrichment', 'blacklist_fraction', 'nucleosome_signal'),
  pt.size = 0.1,
  ncol = 5
)

ggsave("VlnPlot_atac_pbmc_10k.pdf", width=10)

## Filter data with naive criteria
pbmc <- subset(
  x = pbmc,
  subset = nCount_peaks > 3000 &
    nCount_peaks < 20000 &
    FRiP > 0.15 &
    blacklist_fraction < 0.05 &
    nucleosome_signal < 4 &
    TSS.enrichment > 2
)

pbmc <- RunTFIDF(pbmc)
pbmc <- FindTopFeatures(pbmc, min.cutoff = 'q0')
pbmc <- RunSVD(pbmc)

pbmc <- RunUMAP(object = pbmc, reduction = 'lsi', dims = 2:30)
pbmc <- FindNeighbors(object = pbmc, reduction = 'lsi', dims = 2:30)
pbmc <- FindClusters(object = pbmc, verbose = FALSE, algorithm = 3)
DimPlot(object = pbmc, label = TRUE) + NoLegend()
ggsave("cluster.pdf")
```
</details>

Additionally, the script reported a warning which might be related to random number generating in [`future`](https://www.r-bloggers.com/2020/09/future-1-19-1-making-sure-proper-random-numbers-are-produced-in-parallel-processing/), I don't know whether this will raise any issue and should be fixed.

```
Future ('future_lapply-1') unexpectedly generated random
numbers without specifying argument '[future.]seed'. There is a risk that
those random numbers are not statistically sound and the overall results
might be invalid.
```

Thanks again,
obenno

**Reply by timoast** [reply]
> I would like to ask does Signac have a function to conduct "cell-calling" method as 10x, which is described [here](https://support.10xgenomics.com/single-cell-atac/software/pipelines/latest/algorithms/overview#cells)?

No, we just use a simple count threshold chosen manually, in combination with other QC metrics

> Additionally, the script reported a warning which might be related to random number generating 

See https://github.com/satijalab/seurat/issues/3622

**Reply by obenno** [reply]
Thank you for the reply! To avoid any confusion, could you please add some explanation of each column of the `CountFragments()` output dataframe in the function help section?

**Reply by timoast** [reply]
@obenno I've updated the documentation now on the develop branch

**Reply by obenno** [reply]
Thank you very much, that's really helpful!
